src/controllers/rendering/controllerrenderingtransform.cpp

```diff
@@ -0,0 +1,74 @@
+#include "controllers/rendering/controllerrenderingtransform.h"
+
+#include <QImage>
+
+#include "controllers/rendering/controllerrenderingengine.h"
+#include "util/assert.h"
+
+ControllerRenderingTransformFunctionType transformTypeFromString(const QString& type) {
+    if (type.isEmpty()) {
+        return ControllerRenderingTransformFunctionType::NONE;
+    }
+    if (!QString::compare(type, "JAVASCRIPT", Qt::CaseInsensitive)) {
+        return ControllerRenderingTransformFunctionType::JAVASCRIPT;
+    }
+
+    return ControllerRenderingTransformFunctionType::UNSUPPORTED;
+}
+
+bool ControllerRenderingJSTransformFunction::prepare(const ControllerRenderingEngine* parent) {
+    m_function = parent->jsEngine()->evaluate(m_functionSource);
+
+    if (m_function.isError()) {
+        QString errorMessage = m_function.toString();
+        QString line = m_function.property("lineNumber").toString();
+        QString backtrace = m_function.property("stack").toString();
+
+        qWarning() << "The provided controller rendering transformation "
+                      "function has an error:"
+                   << QString("<passed code>:%2: %3").arg(line, errorMessage);
+        return false;
+    }
+
+    if (!m_function.isCallable()) {
+        qWarning() << "The provided controller rendering transformation function isn't callable!";
+        return false;
+    }
+    return true;
+}
+
+bool ControllerRenderingJSTransformFunction::transform(
+        const ControllerRenderingEngine* parent,
+        QByteArray& output,
+        uint8_t screenId) {
+    const QImage frame = parent->fbo()->toImage();
```

Comment:
```suggestion
    const QImage fboImage= parent->fbo()->toImage();
    QImage frame(fboImage.constBits(), fboImage.width(), fboImage.height(), QImage::Format_RGB16);
```
Could you try this. I adapted it from the example in https://doc.qt.io/qt-6/qopenglframebufferobject.html#toImage

Reply:
I have tried that already. I believe the issue comes from the fact that when `toImage` renders then reads the GL output buffer using `glReadPixels`, it also changes the GL format (I had found the source of `toImage` on Qt 5.11 but I guess they haven't changed the behaviour)

Reply:
Maybe @m0dB has an idea, how to render the frame to an RGB565 image?

Reply:
I have never done this, but instead of glReadPixels, which converts the format, I think you should be able to use glGetTexImage to access the data of the FBO's texture in it's original format.